From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hector Martin <marcan@marcan.st>
Date: Tue, 15 Feb 2022 18:45:25 +0900
Subject: [PATCH 02/20] nvmem: Add simple-mfd-nvmem driver

Add driver that exposes part of a simple MFD device as an NVMEM device.
It is intended to be used with e.g. PMICs that are used to hold
power-management configuration, such as used on Apple iPhones.

Signed-off-by: Hector Martin <marcan@marcan.st>
Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 drivers/nvmem/Kconfig            | 12 ++++
 drivers/nvmem/Makefile           |  2 +
 drivers/nvmem/simple-mfd-nvmem.c | 98 ++++++++++++++++++++++++++++++++
 3 files changed, 112 insertions(+)
 create mode 100644 drivers/nvmem/simple-mfd-nvmem.c

diff --git a/drivers/nvmem/Kconfig b/drivers/nvmem/Kconfig
index c36c2a4c2..b9cf44398 100644
--- a/drivers/nvmem/Kconfig
+++ b/drivers/nvmem/Kconfig
@@ -465,6 +465,18 @@ config NVMEM_SNVS_LPGPR
 	  This driver can also be built as a module. If so, the module
 	  will be called nvmem-snvs-lpgpr.
 
+config NVMEM_SIMPLE_MFD
+	tristate "Simple MFD - I2C/SPMI NVMEM"
+	depends on MFD_SIMPLE_MFD_I2C || MFD_SIMPLE_MFD_SPMI || COMPILE_TEST
+	help
+	  Say y here to build a generic driver to expose an simple MFD device
+	  as a NVMEM provider. Both I2C and SPMI MFDs are supported. This can
+	  be used for PMIC devices which are used to store power and
+	  RTC-related settings on certain platforms, such as Apple iPhones.
+
+	  This driver can also be built as a module. If so, the module
+	  will be called nvmem-simple-mfd.
+
 config NVMEM_SPMI_SDAM
 	tristate "SPMI SDAM Support"
 	depends on SPMI
diff --git a/drivers/nvmem/Makefile b/drivers/nvmem/Makefile
index 2bbfb9ff1..56459426c 100644
--- a/drivers/nvmem/Makefile
+++ b/drivers/nvmem/Makefile
@@ -79,6 +79,8 @@ obj-$(CONFIG_NVMEM_SC27XX_EFUSE)	+= nvmem-sc27xx-efuse.o
 nvmem-sc27xx-efuse-y			:= sc27xx-efuse.o
 obj-$(CONFIG_NVMEM_SNVS_LPGPR)		+= nvmem_snvs_lpgpr.o
 nvmem_snvs_lpgpr-y			:= snvs_lpgpr.o
+obj-$(CONFIG_NVMEM_SIMPLE_MFD)		+= nvmem_simple_mfd.o
+nvmem_simple_mfd-y 			:= simple-mfd-nvmem.o
 obj-$(CONFIG_NVMEM_SPMI_SDAM)		+= nvmem_qcom-spmi-sdam.o
 nvmem_qcom-spmi-sdam-y			+= qcom-spmi-sdam.o
 obj-$(CONFIG_NVMEM_SPRD_EFUSE)		+= nvmem_sprd_efuse.o
diff --git a/drivers/nvmem/simple-mfd-nvmem.c b/drivers/nvmem/simple-mfd-nvmem.c
new file mode 100644
index 000000000..2ae862185
--- /dev/null
+++ b/drivers/nvmem/simple-mfd-nvmem.c
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: GPL-2.0-only OR MIT
+/*
+ * Simple MFD - I2C/SPMI NVMEM driver
+ *
+ * Copyright The Asahi Linux Contributors
+ */
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/nvmem-provider.h>
+#include <linux/of.h>
+#include <linux/platform_device.h>
+#include <linux/regmap.h>
+
+struct simple_mfd_nvmem {
+	struct regmap *regmap;
+	unsigned int base;
+};
+
+static int simple_mfd_nvmem_read(void *priv, unsigned int offset,
+                               void *val, size_t bytes)
+{
+	struct simple_mfd_nvmem *nvmem = priv;
+
+        return regmap_bulk_read(nvmem->regmap, nvmem->base + offset, val, bytes);
+}
+
+static int simple_mfd_nvmem_write(void *priv, unsigned int offset,
+                                void *val, size_t bytes)
+{
+	struct simple_mfd_nvmem *nvmem = priv;
+
+	return regmap_bulk_write(nvmem->regmap, nvmem->base + offset, val, bytes);
+}
+
+static int simple_mfd_nvmem_probe(struct platform_device *pdev)
+{
+	struct simple_mfd_nvmem *nvmem;
+	const __be32 *addr;
+	int len;
+	struct nvmem_config nvmem_cfg = {
+		.dev = &pdev->dev,
+		.name = "simple_mfd_nvmem",
+		.id = NVMEM_DEVID_AUTO,
+		.word_size = 1,
+		.stride = 1,
+		.reg_read = simple_mfd_nvmem_read,
+		.reg_write = simple_mfd_nvmem_write,
+		.add_legacy_fixed_of_cells = true,
+	};
+
+	nvmem = devm_kzalloc(&pdev->dev, sizeof(*nvmem), GFP_KERNEL);
+	if (!nvmem)
+		return -ENOMEM;
+
+	nvmem_cfg.priv = nvmem;
+
+	nvmem->regmap = dev_get_regmap(pdev->dev.parent, NULL);
+	if (!nvmem->regmap) {
+		dev_err(&pdev->dev, "Parent regmap unavailable.\n");
+		return -ENXIO;
+	}
+
+	addr = of_get_property(pdev->dev.of_node, "reg", &len);
+	if (!addr) {
+		dev_err(&pdev->dev, "No reg property\n");
+		return -EINVAL;
+	}
+	if (len != 2 * sizeof(u32)) {
+		dev_err(&pdev->dev, "Invalid reg property\n");
+		return -EINVAL;
+	}
+
+	nvmem->base = be32_to_cpup(&addr[0]);
+	nvmem_cfg.size = be32_to_cpup(&addr[1]);
+
+	return PTR_ERR_OR_ZERO(devm_nvmem_register(&pdev->dev, &nvmem_cfg));
+}
+
+static const struct of_device_id simple_mfd_nvmem_id_table[] = {
+	{ .compatible = "apple,pmic-nvmem" },
+	{}
+};
+MODULE_DEVICE_TABLE(of, simple_mfd_nvmem_id_table);
+
+static struct platform_driver simple_mfd_nvmem_driver = {
+	.probe = simple_mfd_nvmem_probe,
+	.driver = {
+		.name = "simple-mfd-nvmem",
+		.of_match_table	= simple_mfd_nvmem_id_table,
+	},
+};
+
+module_platform_driver(simple_mfd_nvmem_driver);
+
+MODULE_LICENSE("Dual MIT/GPL");
+MODULE_AUTHOR("Hector Martin <marcan@marcan.st>");
+MODULE_DESCRIPTION("Simple MFD - I2C/SPMI NVMEM driver");
